app/models/item.rb

```diff
@@ -100,4 +116,63 @@ def persist_revision
       item_revision.save
     end
   end
+
+  def cleanup_revisions_for_a_day(days_from_today, allowed_revisions_count)
+    date = Time.now.utc.to_date - days_from_today
+    revisions_from_date_count = revisions.where(creation_date: date).count
+
+    if revisions_from_date_count > allowed_revisions_count
+      revisions_from_date = revisions
+        .select(:uuid)
+        .where(creation_date: date)
+        .order(creation_date: :desc)
+        .pluck(:uuid)
+
+      revisions_slice_size = (revisions_from_date.length.to_f / allowed_revisions_count).floor
+      revisions_divided_into_slices = revisions_from_date.each_slice(revisions_slice_size).to_a
+      first_slice = revisions_divided_into_slices.shift
+      last_slice = revisions_divided_into_slices.pop
+
+      revisions_to_keep = [
+        first_slice.first,
+        last_slice.last,
+      ]
+
+      beginning_counter = 0
+      end_counter = revisions_divided_into_slices.length - 1
+      counter = 0
+      while revisions_to_keep.length < allowed_revisions_count
+        if counter.odd?
+          revisions_to_keep.push(
+            revisions_divided_into_slices[beginning_counter][
+              (revisions_divided_into_slices[beginning_counter].length.to_f / 2).floor
+            ]
+          )
+          beginning_counter += 1
+        else
+          revisions_to_keep.push(
+            revisions_divided_into_slices[end_counter][
+              (revisions_divided_into_slices[end_counter].length.to_f / 2).floor
+            ]
+          )
+          end_counter -= 1
+        end
+
+        counter += 1
+      end
+
+      Revision
```

Comment:
Could this be changed to reference our item's revisions property instead?

Something like: 
```
self.revisions.where(creation_date: date).where.not(uuid: revisions_to_keep).delete_all
```

Is that essentially the same query? If it is, it might be better to access our item's revisions, since this is an instance method.

Reply:
Also since we already accessed `revisions` earlier in this function call it might already be loaded/cached.

Reply:
Good point - will do 👍

Reply:
Can you run `explain` on this and make sure it's accessing the same key/index we recently created? Accessing self.item_revisions goes through a join table through Rails magic so just don't wanna depend fully on that working as we expect. We can verify it's the same query I think by running explain I believe?

Reply:
Woah ... how Rails does it's "magic" is a disaster ... look closely at the first delete query (debug data from running tests)

<img width="2553" alt="Zrzut ekranu 2020-11-4 o 09 39 08" src="https://user-images.githubusercontent.com/311058/98090163-f54dd580-1e83-11eb-9d2f-ab8d530967e3.png">

`self.revisions...` does not even inner join `item` with `revisions`, this would wipe out the entire revisions table as there is no lock on a specific item in the query. It's high time we started ditching Rails for it's suprising "magic". 

Reverted back the the explicit query on Revisions and ItemRevisions - at least it gives some sense of security on how it behaves and we can utilize the index we already have. Additionally it's a simple delete query (one with inner join would be definitely more complex).

Reply:
Does that mean anytime we access self.revisions it's not accessing the right values?

Reply:
`self.revisions` works proper:
<img width="2559" alt="Zrzut ekranu 2020-11-5 o 08 48 39" src="https://user-images.githubusercontent.com/311058/98212618-268dda80-1f44-11eb-90e3-462454542545.png">

`self.revisions.where(creation_date: date).where.not(uuid: revisions_to_keep).delete_all` does not.

Out of curiosity I've also checked `self.revisions.where(creation_date: date).where.not(uuid: revisions_to_keep).destroy_all` and it just bombs the database with a cascade of select/delete queries (also no inner join).

<img width="2555" alt="Zrzut ekranu 2020-11-5 o 08 51 14" src="https://user-images.githubusercontent.com/311058/98212772-5dfc8700-1f44-11eb-8d1a-275fa76f5cfc.png">